### PR TITLE
Fix p12 creation error when certificate contains auxiliary data

### PR DESCRIPTION
**Pull request by obatysh**
CLA: provided

PKCS12_create_ex method called PKCS12_add_cert (if certificate was specified), which added friendly name and local key id attributes if present in the certificate auxiliary data. Then it tried to add friendly name specified by the user and local key id calculated earlier again. Then error with duplicate attributes happened and p12 was not created.

Added a check not to add friendly name and local key id attributes for a second time if they were already added.

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Comment by obatysh** [comment]
Also happens on 3.1 branch

File: test/pkcs12_format_test.c [hunk 1]
```diff
@@ -882,6 +882,66 @@ static int pkcs12_create_test(void)
...
+    const unsigned char *key_bytes = KEY1;
+    BIO *bio = NULL;
+
+    cert = d2i_X509(NULL, &cert_bytes, sizeof(CERT1));
+    if (!TEST_ptr(cert))
+        goto err;
+    pkey = d2i_AutoPrivateKey(NULL, &key_bytes, sizeof(KEY1));
+    if (!TEST_ptr(pkey))
+        goto err;
+    p12 = PKCS12_create("pass", NULL, pkey, cert, NULL, NID_aes_256_cbc, NID_aes_256_cbc, 2, 1, 0);
```
Comment (t8m) [inline test/pkcs12_format_test.c:905]: Please split the lines longer than 80 characters according to the code style. https://www.openssl.org/policies/technical/coding-style.html

File: crypto/pkcs12/p12_crt.c [hunk 2]
```diff
@@ -64,9 +64,10 @@ PKCS12 *PKCS12_create_ex2(const char *pass, const char *name, EVP_PKEY *pkey,
 
     if (cert) {
         bag = PKCS12_add_cert(&bags, cert);
-        if (name && !PKCS12_add_friendlyname(bag, name, -1))
+        /* PKCS12_add_cert may add friendlyName and localKeyID attributes if present in the certificate, so check before adding them again */
+        if (PKCS12_SAFEBAG_get0_attr(bag, NID_friendlyName) == NULL && name && !PKCS12_add_friendlyname(bag, name, -1))
```
Comment (t8m) [inline crypto/pkcs12/p12_crt.c:68]: As you're touching this, please use `name != NULL`.

However I am concerned if this is the right fix here - should we actually use the passed-in name to override the existing one?

File: crypto/pkcs12/p12_crt.c [hunk 3]
```diff
@@ -64,9 +64,10 @@ PKCS12 *PKCS12_create_ex2(const char *pass, const char *name, EVP_PKEY *pkey,
...
             goto err;
-        if (keyidlen && !PKCS12_add_localkeyid(bag, keyid, keyidlen))
+        if (PKCS12_SAFEBAG_get0_attr(bag, NID_localKeyID) == NULL && keyidlen && !PKCS12_add_localkeyid(bag, keyid, keyidlen))
```
Comment (t8m) [inline crypto/pkcs12/p12_crt.c:70]: Here I assume this fix is correct. However please split the line.

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/pkcs12_format_test.c (hunk 1 above)
Comment (obatysh) [reply test/pkcs12_format_test.c]: Updated

File: crypto/pkcs12/p12_crt.c (hunk 2 above)
Comment (obatysh) [reply crypto/pkcs12/p12_crt.c]: Updated

File: crypto/pkcs12/p12_crt.c (hunk 3 above)
Comment (obatysh) [reply crypto/pkcs12/p12_crt.c]: Updated

**Comment by t8m** [comment]
Note to the person who will be merging - Please squash the commits as the second one is just a fixup.

**Review by tmshort — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Comment by obatysh** [comment]
@tmshort Could you please review the pull request again? The necessary changes were made.

**Comment by obatysh** [comment]
Can this fix be part of the 3.1.3 release?

**Comment by t8m** [comment]
Only if someone from @openssl/committers reviews this ASAP.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Squashed and merged to master, 3.1, and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**